airflow/api_fastapi/execution_api/routes/task_instances.py

```diff
@@ -199,17 +199,22 @@ def ti_update_state(
         )
 
     # We exclude_unset to avoid updating fields that are not set in the payload
-    data = ti_patch_payload.model_dump(exclude_unset=True)
+    # We do not need to deserialize "should_retry" -- it is used for dynamic decision-making within failed state
+    data = ti_patch_payload.model_dump(exclude_unset=True, exclude={"should_retry"})
 
     query = update(TI).where(TI.id == ti_id_str).values(data)
 
     if isinstance(ti_patch_payload, TITerminalStatePayload):
         query = TI.duration_expression_update(ti_patch_payload.end_date, query, session.bind)
         query = query.values(state=ti_patch_payload.state)
+        updated_state = ti_patch_payload.state
         if ti_patch_payload.state == State.FAILED:
             # clear the next_method and next_kwargs
             query = query.values(next_method=None, next_kwargs=None)
-            updated_state = State.FAILED
+            task_instance = session.get(TI, ti_id_str)
+            if _is_eligible_to_retry(task_instance, ti_patch_payload.should_retry):
```

Comment:
Do we need to get entire TI here?

Reply:
You can also get `max_tries` & `try_number` too on L188 and here simply compare

https://github.com/apache/airflow/blob/873e765376eb9654e5c8b8a71d95f92dd96f8958/airflow/api_fastapi/execution_api/routes/task_instances.py#L188

Reply:
No, we do not require the entire TI. Only state, max_tries and try_number.
I would still like to keep the logic inside a function or this part would get bulged with details it doesn't need to have and also its better to have a function with lot of code comments.

Reply:
However, I will get the max_tries and try_number from L188 and pass it down.